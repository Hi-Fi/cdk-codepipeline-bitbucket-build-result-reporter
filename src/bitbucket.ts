import * as AWS from '@aws-sdk/client-ssm';
import fetch from 'node-fetch';

const ssm = new AWS.SSM({});

export const getToken = async (Name: string) => {
  const result = await ssm.getParameter({ Name, WithDecryption: true });
  if (result.Parameter && result.Parameter.Value) {
    return result.Parameter?.Value;
  }
  throw new Error('Cannot fetch token');
};

export interface BitbucketBuildStatus {
  'state': 'INPROGRESS' | 'SUCCESSFUL' | 'FAILED';
  'key': string;
  'name': string;
  'url': string;
  'description': string;
}

// See https://docs.atlassian.com/bitbucket-server/rest/4.0.0/bitbucket-build-rest.html
export const putCodePipelineResultToBitBucket = async (commitId: string, body: string) => {
  if (!process.env.BITBUCKET_TOKEN) {
    throw new Error('Missing BITBUCKET_TOKEN');
  }
  const token = await getToken(process.env.BITBUCKET_TOKEN);

  const bitbucketBuildStatusUrl = `https://${process.env.BITBUCKET_SERVER}/rest/build-status/1.0/commits/${commitId}`;
  const response = await fetch(bitbucketBuildStatusUrl, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body,
  });

  if (response.status === 204) {
    return { status: 'ok' };
  }

  return response.json();
};
