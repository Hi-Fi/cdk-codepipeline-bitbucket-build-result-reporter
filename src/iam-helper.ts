import * as IAM from '@aws-sdk/client-iam';

const iam = new IAM.IAMClient({});

export const getCurrentAccountAlias = async (accountId: string) => {
  try {
    const aliases = await iam.send(new IAM.ListAccountAliasesCommand({}));
    if (aliases.AccountAliases && aliases.AccountAliases.length > 0) {
      return aliases.AccountAliases[0];
    }
    return accountId;
  } catch (err) {
    console.error(err);
    return accountId;
  }
};
