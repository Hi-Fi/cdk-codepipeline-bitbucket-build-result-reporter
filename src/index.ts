import * as ec2 from '@aws-cdk/aws-ec2';
import * as events from '@aws-cdk/aws-events';
import * as targets from '@aws-cdk/aws-events-targets';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as lambda_nodejs from '@aws-cdk/aws-lambda-nodejs';
import * as cdk from '@aws-cdk/core';

const addCodePipelineActionStateChangeEventRule = (scope: cdk.Construct, states: string[], handler: lambda.IFunction) => {
  new events.Rule(scope, 'CodePipelineActionExecutionStateChangeRule', {
    eventPattern: {
      detailType: ['CodePipeline Action Execution State Change'],
      source: ['aws.codepipeline'],
      detail: { state: states },
    },
    targets: [new targets.LambdaFunction(handler)],
  });
};

const addCodeBuildStateChangeEventRule = (scope: cdk.Construct, states: string[], handler: lambda.IFunction) => {
  new events.Rule(scope, 'CodeCodeBuildStateChangeEventRule', {
    eventPattern: {
      detailType: ['CodeBuild Build State Change'],
      source: ['aws.codebuild'],
      detail: { 'build-status': states },
    },
    targets: [new targets.LambdaFunction(handler)],
  });
};

export interface CodePipelineBitbucketBuildResultReporterProps {
  /**
   * The VPC in which to run the status reporter.
   */
  readonly vpc: ec2.VpcAttributes;

  /**
   * Name of the SSM parameter that contains the BitBucket access token.
   * @default BITBUCKET_UPDATE_BUILD_STATUS_TOKEN
   */
  readonly bitbucketTokenName?: string;

  /**
   * The BitBucket server address.
   */
  readonly bitbucketServerAddress: string;
}

/** A construct for reporting CodePipeline build statuses to a BitBucket server using BitBucket REST API.
 * You need to configure SSM parameter BITBUCKET_UPDATE_BUILD_STATUS_TOKEN before using the component.
 */
export class CodePipelineBitbucketBuildResultReporter extends cdk.Construct {
  constructor(scope: cdk.Construct, id: string, props: CodePipelineBitbucketBuildResultReporterProps) {
    super(scope, id);
    const bitbucketTokenName = props.bitbucketTokenName ?? 'BITBUCKET_UPDATE_BUILD_STATUS_TOKEN';
    const vpc = ec2.Vpc.fromVpcAttributes(scope, 'LambdaVpc', props.vpc);
    const codePipelineStatusHandler = new lambda_nodejs.NodejsFunction(scope, 'CodePipelineStatusHandler', {
      vpc,
      runtime: lambda.Runtime.NODEJS_14_X,
      memorySize: 256,
      description: 'Synchronize CodePipeline build statuses to BitBucket',
      environment: {
        BITBUCKET_SERVER: props.bitbucketServerAddress,
        BITBUCKET_TOKEN: bitbucketTokenName,
      },
    });
    codePipelineStatusHandler.role?.addToPrincipalPolicy(new iam.PolicyStatement({
      actions: ['ssm:GetParameter'],
      resources: [`arn:aws:ssm:*:*:parameter/${bitbucketTokenName}`],
    }));
    codePipelineStatusHandler.role?.addToPrincipalPolicy(new iam.PolicyStatement({
      actions: ['codepipeline:GetPipelineExecution', 'codepipeline:GetPipelineState'],
      resources: ['arn:aws:codepipeline:*:*:*'],
    }));

    // https://docs.aws.amazon.com/codepipeline/latest/userguide/detect-state-changes-cloudwatch-events.html
    const codePipelineStates = ['STARTED', 'SUCCEEDED', 'FAILED', 'CANCELED'];
    addCodePipelineActionStateChangeEventRule(scope, codePipelineStates, codePipelineStatusHandler);

    const codeBuildStatusHandler = new lambda_nodejs.NodejsFunction(scope, 'CodeBuildStatusHandler', {
      vpc,
      runtime: lambda.Runtime.NODEJS_12_X,
      memorySize: 256,
      description: 'Synchronize CodePipeline build statuses to BitBucket',
      environment: {
        BITBUCKET_SERVER: props.bitbucketServerAddress,
        BITBUCKET_TOKEN: bitbucketTokenName,
      },
    });
    codeBuildStatusHandler.role?.addToPrincipalPolicy(new iam.PolicyStatement({
      actions: ['ssm:GetParameter'],
      resources: [`arn:aws:ssm:*:*:parameter/${bitbucketTokenName}`],
    }));
    codeBuildStatusHandler.role?.addToPrincipalPolicy(new iam.PolicyStatement({
      actions: ['codebuild:BatchGetBuilds'],
      resources: ['*'],
    }));
    // https://docs.aws.amazon.com/codepipeline/latest/userguide/detect-state-changes-cloudwatch-events.html
    const codeBuildStates = ['IN_PROGRESS', 'SUCCEEDED', 'FAILED', 'STOPPED'];
    addCodeBuildStateChangeEventRule(scope, codeBuildStates, codeBuildStatusHandler);
  }
}
