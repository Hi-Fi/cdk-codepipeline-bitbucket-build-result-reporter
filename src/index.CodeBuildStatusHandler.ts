import type * as AwsLambda from 'aws-lambda';
import { BitbucketBuildStatus, putCodePipelineResultToBitBucket } from './bitbucket';

export const buildBitbucketBuildStatusBody = (
  event: AwsLambda.CodeBuildCloudWatchStateEvent,
  actionStatus: AwsLambda.CodeBuildStateType): BitbucketBuildStatus => {
  const detail = event.detail;
  const state =
    actionStatus === 'IN_PROGRESS' ? 'INPROGRESS' :
      actionStatus === 'SUCCEEDED' ? 'SUCCESSFUL' : 'FAILED';

  // console.log(`${detail['execution-id']} Build status ${state} being reported for ${detail.pipeline} ${detail.stage}-${detail.action}`);

  return {
    state,
    key: `${detail['project-name']}-${detail['build-status']}-${detail['current-phase']}`,
    name: `${detail['project-name']}-${detail['build-status']}`,
    url: `https://${event.region}.console.aws.amazon.com/codesuite/codebuild/459568918122/projects/${detail['project-name']}/build/${detail['project-name']}:${detail['build-id']}/?region=${event.region}`,
    description: `${detail['project-name']} build initiated by ${detail['additional-information'].initiator} at ${detail['additional-information']['build-start-time']}`,
  };
};

exports.handler = async (event: AwsLambda.CodeBuildCloudWatchStateEvent) => {
  // console.log(JSON.stringify(event, undefined, 2));

  try {
    const status = buildBitbucketBuildStatusBody(event, event.detail['build-status']);

    const result = await putCodePipelineResultToBitBucket(event.detail['additional-information']['source-version'], JSON.stringify(status));
    console.log(result);
  } catch (error) {
    console.error(JSON.stringify(error));
  }
};
