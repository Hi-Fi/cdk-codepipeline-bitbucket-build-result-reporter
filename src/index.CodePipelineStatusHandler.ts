import type * as AwsLambda from 'aws-lambda';
import * as AWS from 'aws-sdk';
import { BitbucketBuildStatus, putCodePipelineResultToBitBucket } from './bitbucket';

const codePipeline = new AWS.CodePipeline();

export const buildBitbucketBuildStatusBody = (
  event: AwsLambda.CodePipelineCloudWatchActionEvent,
  actionStatus: AWS.CodePipeline.ActionExecutionStatus): BitbucketBuildStatus => {
  const detail = event.detail;
  const state =
    actionStatus === 'InProgress' ? 'INPROGRESS' :
      actionStatus === 'Succeeded' ? 'SUCCESSFUL' :
        actionStatus === 'Abandoned' ? 'SUCCESSFUL' : 'FAILED';

  console.log(`${detail['execution-id']} Build status ${state} being reported for ${detail.pipeline} ${detail.stage}-${detail.action}`);

  return {
    state,
    key: `${detail.stage}-${detail.action}`,
    name: detail.action,
    url: `https://${event.region}.console.aws.amazon.com/codesuite/codepipeline/pipelines/${detail.pipeline}/view`,
    description: `${detail.stage}-${detail.action}`,
  };
};

const fetchExecution = async (event: AwsLambda.CodePipelineCloudWatchActionEvent) => {
  const detail = event.detail;
  const execution = await codePipeline.getPipelineExecution({
    pipelineName: detail.pipeline,
    pipelineExecutionId: detail['execution-id'],
  }).promise();

  if (!execution.pipelineExecution) {
    throw new Error(`Could not fetch execution ${detail['execution-id']}`);
  }

  return execution.pipelineExecution;
};

const getPipelineActionLatestStatus = async (event: AwsLambda.CodePipelineCloudWatchActionEvent): Promise<AWS.CodePipeline.ActionExecutionStatus> => {
  const pipelineState = await codePipeline.getPipelineState({
    name: event.detail.pipeline,
  }).promise();

  const stageStates = pipelineState.stageStates?.find((s) => s.stageName === event.detail.stage);
  const action = stageStates?.actionStates?.find((a) => a.actionName === event.detail.action);
  return action?.latestExecution?.status ?? 'Failed';
};

exports.handler = async (event: AwsLambda.CodePipelineCloudWatchActionEvent) => {
  // console.log(JSON.stringify(event, undefined, 2));

  try {
    const actionStatus = await getPipelineActionLatestStatus(event);
    const status = buildBitbucketBuildStatusBody(event, actionStatus);

    const pipelineExecution = await fetchExecution(event);
    const revisions = pipelineExecution.artifactRevisions ?? [{ revisionChangeIdentifier: '' }];
    const revision = revisions[0].revisionId ?? '';

    const result = await putCodePipelineResultToBitBucket(revision, JSON.stringify(status));
    console.log(result);
  } catch (error) {
    console.error(JSON.stringify(error));
  }
};
