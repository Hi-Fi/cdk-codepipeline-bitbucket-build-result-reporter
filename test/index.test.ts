import { Capture } from '@aws-cdk/assert';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as cdk from '@aws-cdk/core';
import { CodePipelineBitbucketBuildResultReporter } from '../src/index';
import '@aws-cdk/assert/jest';

test('Create CodePipelineBitbucketBuildResultReporter', () => {
  const mockApp = new cdk.App();
  const stack = new cdk.Stack(mockApp, 'testing-stack');

  const fakeVpc: ec2.VpcAttributes = {
    vpcId: 'vpc-12345678',
    vpcCidrBlock: '10.10.10.0/23',
    availabilityZones: ['eu-central-1a', 'eu-central-1b'],
    publicSubnetIds: ['subnet-1122', 'subnet-2233'],
    publicSubnetRouteTableIds: ['rtb-4444', 'rtb-4444'],
    privateSubnetIds: ['subnet-8888', 'subnet-9999'],
    privateSubnetRouteTableIds: ['rtb-6666', 'rtb-6666'],
  };

  new CodePipelineBitbucketBuildResultReporter(stack, 'CodePipelineBitbucketBuildResultReporter', {
    bitbucketServerAddress: 'bitbucket-server.com',
    bitbucketTokenName: '/my/ssm/variable/BITBUCKET_UPDATE_BUILD_STATUS_TOKEN',
    vpc: fakeVpc,
  });

  const someValue = Capture.anyType();
  expect(stack).toHaveResource('AWS::Lambda::Function', {
    Code: {
      S3Bucket: {
        Ref: someValue.capture(),
      },
      S3Key: {
        'Fn::Join': [
          '',
          [
            {
              'Fn::Select': [
                0,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: someValue.capture(),
                    },
                  ],
                },
              ],
            },
            {
              'Fn::Select': [
                1,
                {
                  'Fn::Split': [
                    '||',
                    {
                      Ref: someValue.capture(),
                    },
                  ],
                },
              ],
            },
          ],
        ],
      },
    },
    Handler: 'index.handler',
    Role: {
      'Fn::GetAtt': [
        'CodePipelineStatusHandlerServiceRole49C6622A',
        'Arn',
      ],
    },
    Runtime: 'nodejs14.x',
    Description: 'Synchronize CodePipeline build statuses to BitBucket',
    MemorySize: 256,
    Environment: {
      Variables: {
        BITBUCKET_SERVER: 'bitbucket-server.com',
        BITBUCKET_TOKEN: '/my/ssm/variable/BITBUCKET_UPDATE_BUILD_STATUS_TOKEN',
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
      },
    },
    VpcConfig: {
      SecurityGroupIds: [
        {
          'Fn::GetAtt': [
            'CodePipelineStatusHandlerSecurityGroup7B4F49AA',
            'GroupId',
          ],
        },
      ],
      SubnetIds: [
        'subnet-8888',
        'subnet-9999',
      ],
    },
  });

  expect(stack).toHaveResource('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRole',
          Effect: 'Allow',
          Principal: {
            Service: 'lambda.amazonaws.com',
          },
        },
      ],
      Version: '2012-10-17',
    },
    ManagedPolicyArns: [
      {
        'Fn::Join': [
          '',
          [
            'arn:',
            {
              Ref: 'AWS::Partition',
            },
            ':iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
          ],
        ],
      },
      {
        'Fn::Join': [
          '',
          [
            'arn:',
            {
              Ref: 'AWS::Partition',
            },
            ':iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole',
          ],
        ],
      },
    ],
  });

  expect(stack).toHaveResource('AWS::IAM::Policy', {
    PolicyDocument: {
      Statement: [
        {
          Action: 'ssm:GetParameter',
          Effect: 'Allow',
          Resource: 'arn:aws:ssm:*:*:parameter//my/ssm/variable/BITBUCKET_UPDATE_BUILD_STATUS_TOKEN',
        },
        {
          Action: [
            'codepipeline:GetPipelineExecution',
            'codepipeline:GetPipelineState',
          ],
          Effect: 'Allow',
          Resource: 'arn:aws:codepipeline:*:*:*',
        },
      ],
      Version: '2012-10-17',
    },
    PolicyName: 'CodePipelineStatusHandlerServiceRoleDefaultPolicy8241AF3D',
    Roles: [
      {
        Ref: 'CodePipelineStatusHandlerServiceRole49C6622A',
      },
    ],
  });

  expect(stack).toHaveResource('AWS::Events::Rule', {
    EventPattern: {
      'detail-type': [
        'CodePipeline Action Execution State Change',
      ],
      'source': [
        'aws.codepipeline',
      ],
      'detail': {
        state: [
          'STARTED',
          'SUCCEEDED',
          'FAILED',
          'CANCELED',
        ],
      },
    },
    State: 'ENABLED',
    Targets: [
      {
        Arn: {
          'Fn::GetAtt': [
            'CodePipelineStatusHandlerA91EA00A',
            'Arn',
          ],
        },
        Id: 'Target0',
      },
    ],
  });

  expect(stack).toHaveResource('AWS::Events::Rule', {
    EventPattern: {
      'detail-type': [
        'CodeBuild Build State Change',
      ],
      'source': [
        'aws.codebuild',
      ],
      'detail': {
        'build-status': [
          'IN_PROGRESS',
          'SUCCEEDED',
          'FAILED',
          'STOPPED',
        ],
      },
    },
    State: 'ENABLED',
    Targets: [
      {
        Arn: {
          'Fn::GetAtt': [
            'CodeBuildStatusHandlerE1754F65',
            'Arn',
          ],
        },
        Id: 'Target0',
      },
    ],
  });
});
