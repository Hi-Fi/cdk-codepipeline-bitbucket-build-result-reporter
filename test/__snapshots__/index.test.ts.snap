// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodePipelineBitbucketBuildResultReporter produces expected template 1`] = `
Object {
  "Resources": Object {
    "CodeBuildStatusHandlerE1754F65": Object {
      "DependsOn": Array [
        "CodeBuildStatusHandlerServiceRoleDefaultPolicy8BCA6254",
        "CodeBuildStatusHandlerServiceRoleF342BBC4",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2af478f2a1d60548b691e8a80705d42c843c6a7e963d3e6198270fb7b7b95abf.zip",
        },
        "Description": "Synchronize CodeBuild build statuses to BitBucket",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BITBUCKET_SERVER": "bitbucket-server.com",
            "BITBUCKET_TOKEN": "/my/ssm/variable/BITBUCKET_UPDATE_BUILD_STATUS_TOKEN",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CodeBuildStatusHandlerServiceRoleF342BBC4",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "CodeBuildStatusHandlerSecurityGroup47948056",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": Array [
            "subnet-8888",
            "subnet-9999",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "CodeBuildStatusHandlerLogRetention760110C5": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CodeBuildStatusHandlerE1754F65",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "CodeBuildStatusHandlerSecurityGroup47948056": Object {
      "Properties": Object {
        "GroupDescription": "Automatic security group for Lambda Function testingstackCodeBuildStatusHandler2993D58A",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc-12345678",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "CodeBuildStatusHandlerServiceRoleDefaultPolicy8BCA6254": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": "arn:aws:ssm:*:*:parameter//my/ssm/variable/BITBUCKET_UPDATE_BUILD_STATUS_TOKEN",
            },
            Object {
              "Action": "iam:ListAccountAliases",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "codebuild:BatchGetBuilds",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CodeBuildStatusHandlerServiceRoleDefaultPolicy8BCA6254",
        "Roles": Array [
          Object {
            "Ref": "CodeBuildStatusHandlerServiceRoleF342BBC4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CodeBuildStatusHandlerServiceRoleF342BBC4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CodeCodeBuildStateChangeEventRule9537577D": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail": Object {
            "build-status": Array [
              "IN_PROGRESS",
              "SUCCEEDED",
              "FAILED",
              "STOPPED",
            ],
          },
          "detail-type": Array [
            "CodeBuild Build State Change",
          ],
          "source": Array [
            "aws.codebuild",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "CodeBuildStatusHandlerE1754F65",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "CodeCodeBuildStateChangeEventRuleAllowEventRuletestingstackCodeBuildStatusHandler2993D58A58E9E0F4": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "CodeBuildStatusHandlerE1754F65",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "CodeCodeBuildStateChangeEventRule9537577D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CodePipelineActionExecutionStateChangeRuleAllowEventRuletestingstackCodePipelineStatusHandler8D6F4B47692CF121": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "CodePipelineStatusHandlerA91EA00A",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "CodePipelineActionExecutionStateChangeRuleBDB3DFB0",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CodePipelineActionExecutionStateChangeRuleBDB3DFB0": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail": Object {
            "state": Array [
              "STARTED",
              "SUCCEEDED",
              "FAILED",
              "CANCELED",
            ],
          },
          "detail-type": Array [
            "CodePipeline Action Execution State Change",
          ],
          "source": Array [
            "aws.codepipeline",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "CodePipelineStatusHandlerA91EA00A",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "CodePipelineStatusHandlerA91EA00A": Object {
      "DependsOn": Array [
        "CodePipelineStatusHandlerServiceRoleDefaultPolicy8241AF3D",
        "CodePipelineStatusHandlerServiceRole49C6622A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "10e16c1d52accb9a2053509c031bcf452f8ca6e7b7b5cc80fcc3826648e4fbbc.zip",
        },
        "Description": "Synchronize CodePipeline build statuses to BitBucket",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BITBUCKET_SERVER": "bitbucket-server.com",
            "BITBUCKET_TOKEN": "/my/ssm/variable/BITBUCKET_UPDATE_BUILD_STATUS_TOKEN",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 256,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CodePipelineStatusHandlerServiceRole49C6622A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "CodePipelineStatusHandlerSecurityGroup7B4F49AA",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": Array [
            "subnet-8888",
            "subnet-9999",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "CodePipelineStatusHandlerLogRetention5118C73F": Object {
      "Properties": Object {
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CodePipelineStatusHandlerA91EA00A",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "CodePipelineStatusHandlerSecurityGroup7B4F49AA": Object {
      "Properties": Object {
        "GroupDescription": "Automatic security group for Lambda Function testingstackCodePipelineStatusHandler8D6F4B47",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": "vpc-12345678",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "CodePipelineStatusHandlerServiceRole49C6622A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CodePipelineStatusHandlerServiceRoleDefaultPolicy8241AF3D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParameter",
              "Effect": "Allow",
              "Resource": "arn:aws:ssm:*:*:parameter//my/ssm/variable/BITBUCKET_UPDATE_BUILD_STATUS_TOKEN",
            },
            Object {
              "Action": "iam:ListAccountAliases",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "codepipeline:GetPipelineExecution",
                "codepipeline:GetPipelineState",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:codepipeline:*:*:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CodePipelineStatusHandlerServiceRoleDefaultPolicy8241AF3D",
        "Roles": Array [
          Object {
            "Ref": "CodePipelineStatusHandlerServiceRole49C6622A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": Object {
      "DependsOn": Array [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f3b4457410e8875dce33608b6046e993c1ba6def4d00e9de1ed9681517a35e45.zip",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": Array [
          Object {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
